amo2kinto/generator.py

```diff
@@ -0,0 +1,121 @@
+import os
+import datetime
+
+from dateutil.parser import parse as dateutil_parser
+from jinja2 import Environment, PackageLoader
+from kinto_http import cli_utils
+
+from . import constants
+from .logger import logger
+
+
+JSON_DATE_FORMAT = "%Y-%m-%dT%H:%M:%SZ"
+COLLECTION_FORMAT = '/buckets/{bucket_id}/collections/{collection_id}'
+
+
+def iso2human(value):
+    date = dateutil_parser(value)
+    return date.strftime('%b %d, %Y')
+
+
+def get_template(name):
+    loader = PackageLoader('amo2kinto', 'templates')
+    env = Environment(loader=loader)
+    env.filters['datetime'] = iso2human
+    return env.get_template(name)
+
+
+def generate_index(records, template, target_dir):
+
+    def sort_records(record):
+        return record['details']['created'], record['last_modified']
+
+    records = sorted(records, key=sort_records, reverse=True)
+    res = template.render(records=records)
+    filename = os.path.join(target_dir, 'index.html')
+
+    logger.info('Writing %s' % filename)
+    with open(filename, 'w') as f:
+        f.write(res.encode('utf8'))
+
+
+def generate_record(record, template, target_dir, get_filename):
+    res = template.render(record=record)
+    filename = os.path.join(target_dir, get_filename(record))
+
+    logger.info('Writing %s' % filename)
+
+    # This needs to be able to write to S3
+    with open(filename, 'w') as f:
+        f.write(res.encode('utf-8'))
+
+
+def get_record_filename(record):
+    return '%s.html' % record.get('blockID', record['id'])
+
+
+def generate(client, collections, target_dir, collection_template, record_template):
+    logger.info('Reading data from Kinto %s...' % client.session.server_url)
+    records = []
+
+    for collection_uri in collections:
+        parts = collection_uri.strip('/').split('/')
+        bucket_id = parts[1]
+        collection_id = parts[3]
+
+        logger.warn('Reading records from %s...' % collection_uri)
+        response = client.get_records(bucket=bucket_id, collection=collection_id)
+        for record in response:
+            record.setdefault('details', {})
```

Comment:
I'm not crazy about deepcopy either. I think I also don't like that we expect the response to be iterable twice. I think what would make me happiest is defining a schema using Colander or Marshmallow or something, and then doing:

```
for raw_record in response:
    record = schema.parse(raw_record)
    records.append(record)
```

But that might be overkill for this project. Maybe using deepcopy is good enough, or leaving it, or even just doing `append` instead of `extend`.